### PR TITLE
Update data section merging

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: Linking.md [hunk 1]
```diff
@@ -224,14 +225,16 @@ imported or defined locally.
...
-wasm globals require modification.
+The output data section is formed, essentially, by concatenating the data
+sections of the input files.  Since the final location in linear memory of
+any give symbol (C global) is not known until link time, all references to
+global addresses generate `R_WEBASSEMBLY_MEMORY_ADDR_*` relocation entries.
+The compiler ensures that each C global is assigned a wasm
+[global](https://github.com/WebAssembly/design/blob/master/Modules.md#global-variables)
+and references to C globals generate relocations referencing the
+corresponding wasm global.  The addresses stored in these wasm globals are
+offsets into linear memory.  In this way the wasm globals act as symbol table
```
Comment (binji) [inline Linking.md]: Is it worth mentioning that the offsets are relative to the object, or do you think that's clear from context?

File: Linking.md [hunk 2]
```diff
@@ -224,14 +225,16 @@ imported or defined locally.
...
-[global](https://github.com/WebAssembly/design/blob/master/Modules.md#global-variables).
-Each C global is assigned a wasm global, and access to C global variables will
-generate a `get_global` followed by a load/store to/from the resulting address.
-The addresses stored in these wasm globals can then be set by the linker when it
-relocates data segments.  For each wasm global that is used in this way an entry
-in the `reloc` of type `R_DATA` is generated so that the linker knows which
-wasm globals require modification.
+The output data section is formed, essentially, by concatenating the data
+sections of the input files.  Since the final location in linear memory of
+any give symbol (C global) is not known until link time, all references to
```
Comment (binji) [inline Linking.md]: Probably should mention how these globals are imported/exported too. (Or is that done elsewhere in the file?)

Comment (dschuff) [inline Linking.md]: any give -> "any given"?

File: Linking.md [hunk 3]
```diff
@@ -224,14 +225,16 @@ imported or defined locally.
...
+global addresses generate `R_WEBASSEMBLY_MEMORY_ADDR_*` relocation entries.
+The compiler ensures that each C global is assigned a wasm
+[global](https://github.com/WebAssembly/design/blob/master/Modules.md#global-variables)
+and references to C globals generate relocations referencing the
```
Comment (dschuff) [inline Linking.md]: It's probably worthwhile to be a little more specific, or providing an example (e.g. what instructions a load of a C global ends up being, and what relocation is emitted).

Comment (sbc100) [reply Linking.md]: This is somewhat covered above where each relocation type is documented.    Should I repeat some of that here?

File: Linking.md (hunk 1 above)
Comment (sbc100) [reply Linking.md]: Done

File: Linking.md (hunk 2 above)
Comment (sbc100) [reply Linking.md]: Done

File: Linking.md [hunk 4]
```diff
@@ -194,11 +195,15 @@ subsection:
...
+globals.  Defined symbols are modeled as exported I32 globals that contain the
+address of the symbol in linear memory.  Undefined globals are modeled as
+imported I32 globals.  These wasm globals are not used at runtime (i.e. there
+are no `get_global/set_global` instructions that reference them) but are instead
+referenced by `R_WEBASSEMBLY_MEMORY_ADDR*` relocation entries.
 
-The global identified in the linking metadata section as the stack pointer is
-handled specially. During linking, only one of these globals is kept, and all
-references to stack-pointer globals are rewritten to use it.
+In the final linked binary all these global are resolved and the only remaining
```
Comment (dschuff) [inline Linking.md]: global -> globals

File: Linking.md (hunk 3 above)
Comment (dschuff) [reply Linking.md]: Hm... yeah, reading it in the whole context I think it's fine, actually.

**Pull request merged**